service/controller/resource/ensurecpcrs/create.go

```diff
@@ -119,6 +122,63 @@ func (r *Resource) createAWSControlPlaneCR(ctx context.Context, cr infrastructur
 }
 
 func (r *Resource) createG8sControlPlaneCR(ctx context.Context, cr infrastructurev1alpha2.AWSCluster, id string) error {
+	// We fetch the operator version of cluster-operator by taking it from the
+	// Tenant Cluster's Cluster CR, since this is reconciled by cluster-operator.
+	// The alternative solution would be to query cluster-service for the release
+	// of the AWSCluster CR, which would imply to create a dependency on
+	// cluster-service which could potentially break the reconciliation in case
+	// cluster-service becomes unavailable. The assumption is that using the
+	// Kubernetes API of the Control Plane is safer as we work against it all the
+	// time anyway.
+	var ov string
+	{
+		cl := &unstructured.Unstructured{}
+		cl.SetAPIVersion("cluster.x-k8s.io/v1alpha2")
+		cl.SetKind("Cluster")
+
+		err := r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: cr.GetName(), Namespace: cr.GetNamespace()}, cl)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ov = cl.GetLabels()[label.ClusterOperatorVersion]
+	}
+
+	cp := &infrastructurev1alpha2.G8sControlPlane{
+		TypeMeta: infrastructurev1alpha2.NewG8sControlPlaneTypeMeta(),
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				annotation.Docs: "https://godoc.org/github.com/giantswarm/apiextensions/pkg/apis/infrastructure/v1alpha2#G8sControlPlane",
+			},
+			Labels: map[string]string{
+				label.ClusterOperatorVersion: ov,
+				label.Cluster:                key.ClusterID(&cr),
+				label.Organization:           key.OrganizationID(&cr),
+				label.Release:                key.ReleaseVersion(&cr),
+			},
+			Name: id,
+		},
+		Spec: infrastructurev1alpha2.G8sControlPlaneSpec{
+			InfrastructureRef: corev1.ObjectReference{
+				APIVersion: infrastructurev1alpha2.NewAWSControlPlaneTypeMeta().APIVersion,
+				Kind:       infrastructurev1alpha2.NewAWSControlPlaneTypeMeta().Kind,
+				Name:       id,
+				Namespace:  cr.GetNamespace(),
+			},
+			// For the migration we pin the replicas to 1 because we only support
+			// running 1 master instance in a Tenant Cluster. As the HA Masters story
+			// advances we will support multi master setups. For the migration process
+			// we have to stick with 1 master. Users can change replica settings later
+			// on once full multi master support is established.
+			Replicas: 1,
+		},
+	}
+
+	_, err := r.k8sClient.G8sClient().InfrastructureV1alpha2().G8sControlPlanes(cr.GetNamespace()).Create(cp)
```

Comment:
Why are we not catching `alreadyExists` errors here?

Reply:
The layout is as such that the existence check is done in `EnsureCreated` where the general outline of the codepath is made. That was just my decision based on taste for the migration here.